parsing/pprintast.ml

```diff
@@ -992,18 +986,25 @@ and module_type ctxt f x =
                 (type_declaration ctxt) td
           | Pwith_modsubst (li, li2) ->
              pp f "module %a :=@ %a" longident_loc li longident_loc li2 in
-        (match l with
-         | [] -> pp f "@[<hov2>%a@]" (module_type ctxt) mt
-         | _ -> pp f "@[<hov2>(%a@ with@ %a)@]"
-                  (module_type ctxt) mt (list with_constraint ~sep:"@ and@ ") l)
+        pp f "@[<hov2>%a@ with@ %a@]"
+          (module_type1 ctxt) mt (list with_constraint ~sep:"@ and@ ") l
+    | _ -> module_type1 ctxt f x
+
+and module_type1 ctxt f x =
+  if x.pmty_attributes <> [] then module_type ctxt f x
+  else match x.pmty_desc with
+    | Pmty_ident li ->
+        pp f "%a" longident_loc li;
+    | Pmty_alias li ->
+        pp f "(module %a)" longident_loc li;
+    | Pmty_signature (s) ->
+        pp f "@[<hv0>@[<hv2>sig@ %a@]@ end@]" (* "@[<hov>sig@ %a@ end@]" *)
+          (list (signature_item ctxt)) s (* FIXME wrong indentation*)
+    | Pmty_with (mt, []) -> module_type1 ctxt f mt
```

Comment:
It is strange that you match the same construction in both functions, maybe remove this `Pmty_with (mt, [])` clause? (It is not wrong, and in theory it could save parentheses in some cases (although I don't think they can actually happen), but I like sticking to the standard pattern.)

Reply:
I did it to save parentheses, but `Pmty_with (mt, [])` can only occur in artificial parsetrees, so I'll remove it.